src/pyvesync/vesyncfan.py

```diff
@@ -2010,13 +2017,437 @@ def set_display(self, mode: bool) -> bool:
+        """Turn display on."""
+        return self.set_display_enabled(True)
+
+    def turn_off_display(self):
+        """Turn display off."""
+        return self.set_display_enabled(False)
+
+    def set_humidity(self, humidity: int) -> bool:
+        """Set target humidity for humidity mode."""
+        if humidity < 30 or humidity > 80:
+            logger.debug("Humidity value must be set between 30 and 80")
+            return False
+        head, body = self.build_api_dict('setTargetHumidity')
+
+        if not head and not body:
+            return False
+
+        body['payload']['data'] = {
+            'targetHumidity': humidity
+        }
+
+        r, _ = Helpers.call_api(
+            '/cloud/v2/deviceManaged/bypassV2',
+            method='post',
+            headers=head,
+            json_object=body,
+        )
+
+        if r is not None and Helpers.code_check(r):
+            return True
+        logger.debug('Error setting humidity')
+        return False
+
+    def set_humidity_mode(self, mode: str) -> bool:
+        """Set humidifier mode."""
+        if mode not in self.mist_modes:
+            logger.debug('Invalid humidity mode used - %s',
+                         mode)
+            logger.debug('Proper modes for this device are - %s',
+                         str(self.mist_modes))
+            return False
+        head, body = self.build_api_dict('setHumidityMode')
+        if not head and not body:
+            return False
+        body['payload']['data'] = {
+            'workMode': 'autoPro' if mode == 'auto' else mode
+        }
+
+        r, _ = Helpers.call_api(
+            '/cloud/v2/deviceManaged/bypassV2',
+            method='post',
+            headers=head,
+            json_object=body,
+        )
+
+        if r is not None and Helpers.code_check(r):
+            return True
+        logger.debug('Error setting humidity mode')
+        return False
+
+    def set_auto_mode(self) -> bool:
+        """Set humidity mode to auto."""
+        return self.set_humidity_mode('auto')
+
+    def set_manual_mode(self) -> bool:
+        """Set humidity mode to manual."""
+        return self.set_humidity_mode('manual')
+
+    def automatic_stop_on(self) -> bool:
+        """Set humidity mode to auto."""
+        return self.set_humidity_mode('auto')
+
+    def automatic_stop_off(self) -> bool:
+        """Set humidity mode to manual."""
+        return self.set_humidity_mode('manual')
+
+    def set_mist_level(self, level) -> bool:
+        """Set humidifier mist level with int between 0 - 9."""
+        try:
+            level = int(level)
+        except ValueError:
+            level = str(level)
+        if level not in self.mist_levels:
+            logger.debug('Humidifier mist level must be between 0 and 9')
+            return False
+
+        head, body = self.build_api_dict('setVirtualLevel')
+        if not head and not body:
+            return False
+
+        body['payload']['data'] = {
```

Comment:
Hey @tkamenick, how did you find this JSON structure? I write a similar plugin for HomeKit and haven't been able to sniff network traffic from the VeSync app for a while. Was happy to see the missing payload in this PR!

Reply:
@pschroeder89: a lot of trial and error. I had an old rooted Android device in a drawer, so I started with that. Then installed Magisk + Riru + EdXposed + TrustMeAlready. 

On my MacBook: installed Charles proxy + enabled SSL proxying. Followed instructions to install / trust cert on the android device.  

Then, used ProxyDroid on the phone to proxy all traffic through Charles on my laptop. This allowed me to inspect all browser and most app based SSL traffic. I celebrated only to find that the vesync app was still pinned.

To unpin the vesync app, ran frida_server on the phone as root and ran frida client on my macbook with the frida-multiple-unpinning script targeting the vesync app on the phone connected via usb.

A lot of work, but once it's set up you can inspect all SSL traffic from the android device.

Reply:
Ah frida_server is the piece I was missing to bypass pinning. VeSync only started pinning in late 2022, prior to that Charles would work fine. Guess I'll be picking up a used Android phone for debugging.
Thank you!